File: vrpc/riot/local_auth.py
from __future__ import annotations

import logging
import warnings

import requests
import urllib3

from ..constants import (
    CLIENT_PLATFORM,
    FALLBACK_CLIENT_VERSION,
    VALORANT_API,
    riot_lockfile_path,
)

warnings.simplefilter("ignore", urllib3.exceptions.InsecureRequestWarning)

logger = logging.getLogger(__name__)


class RiotNotRunning(Exception):
    pass


def read_lockfile() -> dict:
    path = riot_lockfile_path()
    if not path.exists():
        raise RiotNotRunning("lockfile not found")
    try:
        parts = path.read_text(encoding="utf-8").strip().split(":")
        name, pid, port, password, protocol = parts
        return {"name": name, "pid": pid, "port": port, "password": password, "protocol": protocol}
    except Exception as e:
        raise RiotNotRunning(f"lockfile read error: {e}") from e


class LocalAuth:
    def __init__(self) -> None:
        self.session = requests.Session()
        self.session.verify = False
        self.port: str | None = None
        self.password: str | None = None
        self.protocol: str = "https"
        self.access_token: str | None = None
        self.entitlement_token: str | None = None
        self.puuid: str | None = None
        self.client_version: str | None = None

    @property
    def local_base(self) -> str:
        return f"{self.protocol}://127.0.0.1:{self.port}"

    @property
    def local_auth(self) -> tuple[str, str]:
        return ("riot", self.password or "")

    def local_get(self, path: str, timeout: float = 5.0) -> requests.Response:
        return self.session.get(self.local_base + path, auth=self.local_auth, timeout=timeout)

    def refresh(self) -> bool:
        lock = read_lockfile()
        self.port = lock["port"]
        self.password = lock["password"]
        self.protocol = lock["protocol"]

        try:
            r = self.local_get("/entitlements/v1/token")
        except requests.exceptions.RequestException as e:
            raise RiotNotRunning(f"local API unreachable: {e}") from e

        if r.status_code != 200:
            raise RiotNotRunning(f"entitlements HTTP {r.status_code}")

        data = r.json()
        self.access_token = data.get("accessToken")
        self.entitlement_token = data.get("token")
        self.puuid = data.get("subject")
        if not (self.access_token and self.entitlement_token and self.puuid):
            raise RiotNotRunning("incomplete entitlements response")

        if not self.client_version:
            self.client_version = self._fetch_client_version()
        return True

    def _fetch_client_version(self) -> str:
        try:
            r = requests.get(f"{VALORANT_API}/version", timeout=5)
            if r.status_code == 200:
                return r.json()["data"]["riotClientVersion"]
        except Exception as e:
            logger.debug("ClientVersion fetch failed: %s", e)
        return FALLBACK_CLIENT_VERSION

    def pd_glz_headers(self) -> dict:
        return {
            "Authorization": f"Bearer {self.access_token}",
            "X-Riot-Entitlements-JWT": self.entitlement_token or "",
            "X-Riot-ClientPlatform": CLIENT_PLATFORM,
            "X-Riot-ClientVersion": self.client_version or FALLBACK_CLIENT_VERSION,
        }

    def is_valid(self) -> bool:
        return bool(self.access_token and self.entitlement_token and self.puuid)

File: electron/riot/localAuth.ts
import { existsSync, readFileSync } from 'fs'
import type { AxiosResponse } from 'axios'
import { riotClient, webClient } from '../lib/http'
import {
  CLIENT_PLATFORM,
  FALLBACK_CLIENT_VERSION,
  VALORANT_API,
  riotLockfilePath
} from '../constants'

export class RiotNotRunning extends Error {}

interface Lock {
  name: string
  pid: string
  port: string
  password: string
  protocol: string
}

function readLockfile(): Lock {
  const path = riotLockfilePath()
  if (!existsSync(path)) throw new RiotNotRunning('lockfile not found')
  try {
    const parts = readFileSync(path, 'utf-8').trim().split(':')
    const [name, pid, port, password, protocol] = parts
    return { name, pid, port, password, protocol }
  } catch (e) {
    throw new RiotNotRunning(`lockfile read error: ${String(e)}`)
  }
}


export class LocalAuth {
  port: string | null = null
  password: string | null = null
  protocol = 'https'
  accessToken: string | null = null
  entitlementToken: string | null = null
  puuid: string | null = null
  clientVersion: string | null = null

  get localBase(): string {
    return `${this.protocol}://127.0.0.1:${this.port}`
  }

  private basicAuth(): string {
    return 'Basic ' + Buffer.from(`riot:${this.password ?? ''}`).toString('base64')
  }

  localGet(path: string, timeout = 5000): Promise<AxiosResponse> {
    return riotClient.get(this.localBase + path, {
      headers: { Authorization: this.basicAuth() },
      timeout
    })
  }

  async refresh(): Promise<boolean> {
    const lock = readLockfile()
    this.port = lock.port
    this.password = lock.password
    this.protocol = lock.protocol

    let r: AxiosResponse
    try {
      r = await this.localGet('/entitlements/v1/token')
    } catch (e) {
      throw new RiotNotRunning(`local API unreachable: ${String(e)}`)
    }
    if (r.status !== 200) throw new RiotNotRunning(`entitlements HTTP ${r.status}`)

    const data = r.data ?? {}
    this.accessToken = data.accessToken ?? null
    this.entitlementToken = data.token ?? null
    this.puuid = data.subject ?? null
    if (!this.accessToken || !this.entitlementToken || !this.puuid) {
      throw new RiotNotRunning('incomplete entitlements response')
    }
    if (!this.clientVersion) this.clientVersion = await this.fetchClientVersion()
    return true
  }

  private async fetchClientVersion(): Promise<string> {
    try {
      const r = await webClient.get(`${VALORANT_API}/version`)
      if (r.status === 200) return r.data.data.riotClientVersion
    } catch {
      
    }
    return FALLBACK_CLIENT_VERSION
  }

  pdGlzHeaders(): Record<string, string> {
    return {
      Authorization: `Bearer ${this.accessToken}`,
      'X-Riot-Entitlements-JWT': this.entitlementToken ?? '',
      'X-Riot-ClientPlatform': CLIENT_PLATFORM,
      'X-Riot-ClientVersion': this.clientVersion ?? FALLBACK_CLIENT_VERSION
    }
  }

  isValid(): boolean {
    return !!(this.accessToken && this.entitlementToken && this.puuid)
  }
}
